src/main/java/net/neoforged/neoforge/event/BlockEntityTypeAddBlocksEvent.java

```diff
@@ -0,0 +1,122 @@
+import java.util.Map;
+import java.util.Set;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.SignBlock;
+import net.minecraft.world.level.block.StandingSignBlock;
+import net.minecraft.world.level.block.WallSignBlock;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.neoforged.bus.api.Event;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.event.IModBusEvent;
+import org.jetbrains.annotations.Nullable;
+
+public class BlockEntityTypeAddBlocksEvent extends Event implements IModBusEvent {
+    private static final MethodHandle VALID_BLOCKS_SETTER_METHOD_HANDLE;
+    static {
+        try {
+            VALID_BLOCKS_SETTER_METHOD_HANDLE = MethodHandles.privateLookupIn(BlockEntityType.class, MethodHandles.lookup()).findSetter(BlockEntityType.class, "validBlocks", Set.class);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey;
+    private final BlockEntityType<?> blockEntityType;
+    private final Set<Block> currentValidBlocks;
+    private final Supplier<? extends @Nullable Class<?>> baseClass = Suppliers.memoize(this::getCommonSuperClassForExistingValidBlocks);
+
+    private BlockEntityTypeAddBlocksEvent(ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey, BlockEntityType<?> blockEntityType) {
+        this.blockEntityTypeResourceKey = blockEntityTypeResourceKey;
+        this.blockEntityType = blockEntityType;
+        this.currentValidBlocks = new HashSet<>(blockEntityType.getValidBlocks());
+    }
+
+    /**
+     * The {@link ResourceKey} that this current {@link BlockEntityType} was registered under.
+     */
+    public ResourceKey<BlockEntityType<?>> getBlockEntityTypeResourceKey() {
+        return blockEntityTypeResourceKey;
+    }
+
+    /**
+     * The current {@link BlockEntityType}.
+     */
+    public BlockEntityType<?> getBlockEntityType() {
+        return blockEntityType;
+    }
+
+    /**
+     * An immutable list of the list of valid blocks that will be assigned to replace {@link BlockEntityType#validBlocks}'s value.
+     */
+    public Set<Block> getCurrentValidBlocks() {
+        return Collections.unmodifiableSet(this.currentValidBlocks);
+    }
+
+    /**
+     * Will add the given block to the list of valid blocks we will assign to the current {@link BlockEntityType}.
+     * Make sure the given block is the same or derived from the highest common class of all the entries in valid blocks.
+     * <p>
+     * Example: If the valid blocks list has {@link StandingSignBlock} entry and {@link WallSignBlock} entry, the common class is SignBlock
+     * and the given block must be a {@link SignBlock} or has {@link SignBlock} as a child class in its hierarchy.
+     * </p>
```

Comment:
This closing tag is unnecessary, and might mess with the javadoc inserted ones though I'm unsure; I'd just drop it